### DISCUSSION TITLE
Trouble implementing HSM

### DISCUSSION THREAD

**Discussion by felipefl142** [kiishor/UML-State-Machine-in-C#33] [Q&A] [answered]
Hello, Kiishor!

First, I'd like to congratulate and thank you for your UML State Machine project, it's a miracle for working with embedded systems.

Second, I'm having some issues trying to implement the framework on a STM8S MCU system that I'm working on. I'm using ST Visual Develop IDE on Windows 10 and the thing is, the only C language version that the enviroment allows is C90 (ANSI C) so I keep on getting some errors. One of them is about where the constant expression declarations are made and the other one is about the variable array size that I need to use in the framework for my UML Statechart.  I read that those two are limitations that this C version imposes and I wanted to know how I can deal with these problems. Maybe changing my development enviroment could fix these issues by letting me use a newer C, like C99 but I'm not certain.
Another thing is, as I'm using Windows, I can't use the Posix Threads library like I should, so that could force me to change my enviroment completely, using Linux could be a solution to all of this, but it could take much time and the other programmers wouldn't be very eager to do it. At the moment, I really find myself in a huge dilemma.

![errors](https://user-images.githubusercontent.com/71197747/119682382-39e99000-be19-11eb-9c00-f035d1d6eb45.png)

About ANSI C limitations:
https://stackoverflow.com/questions/8474100/where-you-can-and-cannot-declare-new-variables-in-c

A possible workaround the varible array size problem:
https://stackoverflow.com/questions/4352768/how-do-i-declare-an-array-of-undefined-or-no-initial-size

**Comment by kmilo17pet** [comment]
hi guys, 
At some point I ran into similar problems with some MCUs.
At the end, this proyect  among others like [QPC ](https://github.com/QuantumLeaps) from Miro Samek inspired the development of my own hierarchical state machine interface and it is now included as a module in my cooperative RTOS [QuarkTS](https://github.com/kmilo17pet/QuarkTS)
I hope these problems can be solved in the future. I'd love to see what's new with this project :).
Best

**Comment by kiishor** [comment, marked as answer]
Hello felipefl142,

Thanks for your feedback.
Yes, both the errors you mentioned will be fixed if you use C99 or a newer standard. I hope your development environment supports the newer C standard. I strongly advise you to switch to a newer C standard.

If you are not able to use C99 or a newer standard, then 
1. Disable variable-length array
 You can disable VLA by defining HSM_USE_VARIABLE_LENGTH_ARRAY to 0. then you need to also define MAX_HIERARCHICAL_LEVEL.
MAX_HIERARCHICAL_LEVEL should be the highest hierarchical nesting level among all the state machines in your project. This is explained in my readme file.
https://github.com/kiishor/UML-State-Machine-in-C#disable-variable-length-array
This you can do without modifying framework source files. It is explained in the Configuration section of the above readme file.

2. Compilation error due to variable names not declared at the start of a function. in this case, you need to move variable declaration at the start of the function in hsm.c file. There is no alternative to this problem.

However, I could not understand the issue related to windows that you can't use the Posix thread library. I'm not sure how it is related to my state machine framework.

Best regards,
Nandkishor

**Reply by felipefl142** [reply]
Thanks for answer, Kiishor! Hope you're having a nice day.

So... I disabled VLA and set Max Hierarchical Level to 2 and it solved the 1st problem. But in the 2nd problem, I tried to move variable declarations to the start of the function and still had the same error, at the same line where it was declared before I moved it. (see image 1 below)

![errors2](https://user-images.githubusercontent.com/71197747/120186631-1f812f00-c1ea-11eb-8134-e0247ef19891.png)
 Image 1

And the thing about the pThread library is that the compiler can't include the library so it's functions are undefined (see Image 2 below). 

![errors3](https://user-images.githubusercontent.com/71197747/120186736-417ab180-c1ea-11eb-87d2-79aa388e7171.png)
Image 2

For what it's worth, I read that there is no pThread library to use with Windows. Maybe a win32 version, but I don't even know how to make it work in this occasion. At this point, I don't even know what to do.... ;-;

Best Regards,
Felipe F.

**Reply by kiishor** [reply]
Hi Felipe,

I have created a feature branch (feature/asnsi_C90) to support C90 compiler. I have compiled it using GCC 8.1.0 with C90 mode. You can try to use it.

Still, if you face errors then share your hsm.c file and its output error log file. I can have a look at it.

Regarding your second screenshot: **pTarget_Path undefined**
This error is **not** related to pThread library.  The compiler is not able to find the declaration of *pTarget_Path* variable. It can be due to it is not declared at the start of the function.

`const state_t* pTarget_Path[MAX_HIERARCHICAL_LEVEL];     /* Array to store the target node path */`

Regards,
Nandkishor

**Reply by felipefl142** [reply]
Well, the problems persist. Downloaded the new branch and changed the files, but nothing changed. And about the pTarged_Path, the variable is already declared exactly like you showed but I still get the same error :/

Thanks in advance,
Felipe F.

**Reply by kmilo17pet** [reply]
git clone https://github.com/kiishor/UML-State-Machine-in-C.git -b feature/asnsi_C90

**Reply by kiishor** [reply]
Hi Felipe,

I have compiled hsm.c file with CXSTM8 compiler and fixed all the errors. Please pull the latest commit from (feature/asnsi_C90) branch. Also, I have created a test workspace at "demo\hierarchical_state_machine\build\stm8" location for your reference.

Now this shall solve all your problem related to hsm.c. :)

Regards,
Nandkishor

**Reply by felipefl142** [reply]
Hi Nandkishor! Thanks a lot, it seems like it works now. I compiled the demo and ran on my MCU, just need to inherit my previosly written code.
I will update what happens next, but it is looking good! Again, thanks for your help and congratulations on such an amazing and helpful framework, you're the best!
Looking forward to make this work.

Regards,
Felipe F.

**Reply by felipefl142** [reply]
Hello again,
so... Everything works perfect in respect to hsm.c, just as you said. But now my state machine code (the "demo.c" and it's header, in my case they're named "hmi_sm.x") are providing a total of 247 errors (see img 1 below), and I can't even fathom why. 

![errors4](https://user-images.githubusercontent.com/71197747/121352777-919cf680-c903-11eb-8ce2-86b1b564e5bf.png)
Image 1

Best Regards,
Felipe F.

**Comment by felipefl142** [comment]
Hello again,
so... Everything works perfect in respect to hsm.c, just as you said. But now my state machine code (the "demo.c" and it's header, in my case they're named "hmi_sm.x") are providing a total of 247 errors (see img 1 below), and I can't even fathom why.




![errors4](https://user-images.githubusercontent.com/71197747/123960081-46678800-d985-11eb-9c29-e294f99a43b1.png)

Image 1

Best Regards,
Felipe F.

**Reply by kiishor** [reply]
Hi Felipe,

It is difficult to guess why the compiler is throwing that many errors based on the error screenshot. Last time I resolved these errors in hsm.c by setting CXSTM8 toolchain on my PC. Now I don't have that toolchain on my pc. It is really time-consuming for me to do it again.

You can try to follow the below steps,
1. Comment all the code in demo.c and demo.h.  You should get zero errors now.
2. Then start uncommenting code from the top of demo.h till you get the first error. Make sure that you get only one error at a time.
3. Try to fix that error. If you can't figure out why the compiler is complaining, then **try to rewrite that piece of code in your way**.

I can see that some of the errors are repeated. like `Incomplete declaration` and `misplaced local declaration`. Hence if you solve it in one place you shall be able to apply the same fix in all other places.

Regards,
Nandkishor

**Comment by felipefl142** [comment]
Hi, it seems like it works now. I just need one thing, to associate InfraRed Remote commands with the events. I already have a interrupt routine to receive the commands (4 bytes in total). Just need to know where/how do I make the IR interrupt to trigger my enumerated events. Can you help me?
